File: PatronesEstructurales/flyweight.py
class TipoComida:
    def __init__(self, nombre, imagen, sonido):
        self.nombre = nombre
        self.imagen = imagen
        self.sonido = sonido

    def mostrar(self, mesa, lugar):
        print(f"🍽️ Servir {self.nombre} en mesa {mesa}, lugar {lugar} "
              f"con imagen {self.imagen} y sonido {self.sonido}")


class FabricaComida:
    _tipos = {}

    @staticmethod
    def obtener_tipo(nombre, imagen, sonido):
        if nombre not in FabricaComida._tipos:
            FabricaComida._tipos[nombre] = TipoComida(nombre, imagen, sonido)
        return FabricaComida._tipos[nombre]


class Plato:
    def __init__(self, tipo, mesa, lugar):
        self.tipo = tipo
        self.mesa = mesa
        self.lugar = lugar

    def servir(self):
        self.tipo.mostrar(self.mesa, self.lugar)


# --- Uso ---
tipo_pizza = FabricaComida.obtener_tipo("Pizza", "pizza.png", "croc.mp3")
tipo_burger = FabricaComida.obtener_tipo("Hamburguesa", "burger.png", "crunch.mp3")

ordenes = [
    Plato(tipo_pizza, 1, "A"),
    Plato(tipo_pizza, 1, "B"),
    Plato(tipo_burger, 2, "C"),
    Plato(tipo_burger, 2, "D"),
]

for plato in ordenes:
    plato.servir()

File: PatronesEstructurales/Flyweight.java
package PatronesEstructurales;
import java.util.*;


final class FoodType {
    private final String nombre;
    private final String imagen;
    private final String sonido;

    public FoodType(String nombre, String imagen, String sonido) {
        this.nombre = nombre;
        this.imagen = imagen;
        this.sonido = sonido;
    }

    public void mostrar(int mesa, String lugar) {
        System.out.printf("Servir %s en mesa %d, lugar %s con imagen %s y sonido %s%n",
                nombre, mesa, lugar, imagen, sonido);
    }
}


final class FoodFactory {
    private static final Map<String, FoodType> tipos = new HashMap<>();

    
    public static FoodType obtenerTipo(String nombre, String imagen, String sonido) {
        return tipos.computeIfAbsent(nombre, n -> new FoodType(n, imagen, sonido));
    }

    public static int cacheSize() {
        return tipos.size();
    }
}


final class Dish {
    private final FoodType tipo; 
    private final int mesa;      
    private final String lugar;  

    public Dish(FoodType tipo, int mesa, String lugar) {
        this.tipo = tipo;
        this.mesa = mesa;
        this.lugar = lugar;
    }

    public void servir() {
        tipo.mostrar(mesa, lugar);
    }
}

public class Flyweight{
    public static void main(String[] args) {
        FoodType pizza = FoodFactory.obtenerTipo("Pizza", "pizza.png", "croc.mp3");
        FoodType burger = FoodFactory.obtenerTipo("Hamburguesa", "burger.png", "crunch.mp3");

        List<Dish> ordenes = List.of(
                new Dish(pizza, 1, "A"),
                new Dish(pizza, 1, "B"),
                new Dish(burger, 2, "C"),
                new Dish(burger, 2, "D")
        );

        ordenes.forEach(Dish::servir);

        System.out.println("Tipos en caché (flyweights): " + FoodFactory.cacheSize());
    }
}
